main.py

```diff
@@ -51,62 +149,72 @@ def __init__(self, dir: str, compiler: Compiler, scorer: Scorer, source_file: st
 
-    @functools.lru_cache(maxsize=1024)
-    def _get_randomizer(self, source: str) -> Randomizer:
-        ast = self.parser.parse(source)
-        return Randomizer(ast, self.random)
-
-    def score_base(self) -> Tuple[str, int, str]:
+    def create_and_score_base(self) -> Candiate:
         base_source = perm.perm_evaluate_one(self.permutations)
-
-        # Normalize the C code by e.g. stripping whitespace and pragmas
-        randomizer = self._get_randomizer(base_source)
-        base_source = randomizer.get_current_source()
-
-        start_o = self.compiler.compile(base_source)
-        if start_o is None:
+        base_cand = Candiate.from_source(base_source, self.parser)
+        if not base_cand.compile(self.compiler):
             raise Exception(f"Unable to compile {self.source_file}")
+        base_cand.score(self.scorer)
+        return base_cand
 
-        return (base_source, *self.scorer.score(start_o))
-
-    def permutate_next(self) -> bool:
-        cand_c = next(self.iterator, None)
+    def get_next_seed(self) -> int:
+        if self.permutations.is_random():
+            max_ran = max([self.permutations.perm_count, 10**9])
+            seed = self.random.randint(0, max_ran)
+        else:
+            seed = next(self._seed_iterator, None)
+            
+        return seed
+
+    def eval_candidate(self, seed: int) -> Tuple[Candiate, Profiler]:
+        cand_c = self.permutations.evaluate(seed, EvalState())
         if cand_c is None:
-            return False
+            return None
+
+        t0 = time.time()
+
+        cand = Candiate.from_source(cand_c, self.parser, seed)
 
-        randomizer = self._get_randomizer(cand_c)
-        if self.random.uniform(0, 1) >= RANDOMIZER_KEEP_PROB:
-            randomizer.reset()
         if self.permutations.is_random():
-            randomizer.randomize()
-        self.cur_cand = randomizer.get_current_source()
-        self.iteration += 1
-        return True
+            cand.randomize_ast(self.randomizer)
 
-    def get_source(self) -> str:
-        return self.cur_cand
+        t1 = time.time()
+
+        comp_success = cand.compile(self.compiler)
+
+        t2 = time.time()
+
+        if comp_success:
+            cand.score(self.scorer)
+            new_score, new_hash = cand.score_value, cand.score_hash
+        else:
+            new_hash = None
+            new_score = self.scorer.PENALTY_INF
 
-    def compile(self) -> Optional[str]:
-        return self.compiler.compile(self.get_source())
+        t3 = time.time()
 
-    def score(self, cand_o: str) -> Tuple[int, str]:
-        return self.scorer.score(cand_o)
+        profiler: Profiler = Profiler()
+        profiler.add_stat(Profiler.StatType.perm, t1 - t0)
+        profiler.add_stat(Profiler.StatType.compile, t2 - t1)
+        profiler.add_stat(Profiler.StatType.score, t3 - t2)
+        timings = profiler.get_str_stats()
```

Comment:
creating a new profiler every time makes the results jitter, ideally we should be reusing a single one instead

Reply:
I'd agree, but then we'd have to do some nasty cross-process communication. The profiling has to happen separately on each new thread, and somehow needs to get joined back up. Reusing the profiler class seems the best way?